File: src/ngd_ngg_demo/algorithm_client.py
from __future__ import annotations

import json
import urllib.error
import urllib.parse
import urllib.request
from typing import Any


class AlgorithmClientError(RuntimeError):
    def __init__(self, status: int, message: str):
        super().__init__(f"Algorithm API returned HTTP {status}: {message}")
        self.status = status


class AlgorithmClient:
    def __init__(self, base_url: str, timeout_seconds: float = 5.0) -> None:
        self.base_url = base_url.rstrip("/")
        self.timeout_seconds = timeout_seconds

    def request(
        self, method: str, path: str, body: dict[str, Any] | None = None
    ) -> dict[str, Any]:
        data = None if body is None else json.dumps(body).encode("utf-8")
        request = urllib.request.Request(
            self.base_url + path,
            data=data,
            method=method,
            headers={"Accept": "application/json", "Content-Type": "application/json"},
        )
        try:
            with urllib.request.urlopen(
                request, timeout=self.timeout_seconds
            ) as response:
                raw = response.read()
                return json.loads(raw) if raw else {}
        except urllib.error.HTTPError as exc:
            message = exc.read().decode(errors="replace")
            raise AlgorithmClientError(exc.code, message) from exc
        except urllib.error.URLError as exc:
            raise AlgorithmClientError(503, str(exc.reason)) from exc

    def sync_static_snapshot(
        self, snapshot_id: str, snapshot: dict[str, Any]
    ) -> dict[str, Any]:
        quoted = urllib.parse.quote(snapshot_id, safe=":")
        return self.request(
            "PUT", f"/internal/v1/node-static-snapshots/{quoted}", snapshot
        )

    def calculate(self, request: dict[str, Any]) -> dict[str, Any]:
        return self.request("POST", "/api/v1/node-groups/calculate", request)

File: prc/internal/controller/algorithm_client.go
package controller

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"time"
)

type AlgorithmClient struct {
	BaseURL string
	Client  *http.Client
}

type StaticAck struct {
	AlgorithmBootID  string `json:"algorithmBootId"`
	AcceptedSnapshot string `json:"acceptedSnapshotId"`
	NodeCount        int    `json:"nodeCount"`
}

type CandidateNode struct {
	NodeUID  string `json:"nodeUID"`
	NodeName string `json:"nodeName"`
	Score    int64  `json:"score"`
}

type CandidateGroup struct {
	Rank          int64           `json:"rank"`
	GroupID       string          `json:"groupId"`
	TopologyLevel string          `json:"topologyLevel"`
	GroupScore    float64         `json:"groupScore"`
	Nodes         []CandidateNode `json:"nodes"`
}

type AlgorithmResponse struct {
	RequestID                string           `json:"requestId"`
	TaskUID                  string           `json:"taskUID"`
	NGDUID                   string           `json:"ngdUID"`
	NGDGeneration            int64            `json:"ngdGeneration"`
	AlgorithmBootID          string           `json:"algorithmBootId"`
	NodeStaticSnapshotID     string           `json:"nodeStaticSnapshotId"`
	SchedulerStateSnapshotID string           `json:"schedulerStateSnapshotId"`
	MetricSnapshotID         string           `json:"metricSnapshotId"`
	Status                   string           `json:"status"`
	CandidateNodeGroups      []CandidateGroup `json:"candidateNodeGroups"`
}

func (a AlgorithmClient) putStatic(ctx context.Context, snapshotID string, body any) (StaticAck, error) {
	var result StaticAck
	err := a.do(ctx, http.MethodPut, "/internal/v1/node-static-snapshots/"+url.PathEscape(snapshotID), body, &result)
	return result, err
}

func (a AlgorithmClient) calculate(ctx context.Context, body any, timeout time.Duration) (AlgorithmResponse, error) {
	var result AlgorithmResponse
	requestCtx, cancel := context.WithTimeout(ctx, timeout)
	defer cancel()
	err := a.do(requestCtx, http.MethodPost, "/api/v1/node-groups/calculate", body, &result)
	return result, err
}

func (a AlgorithmClient) do(ctx context.Context, method, path string, body, result any) error {
	raw, err := json.Marshal(body)
	if err != nil {
		return fmt.Errorf("marshal Algorithm request: %w", err)
	}
	request, err := http.NewRequestWithContext(ctx, method, a.BaseURL+path, bytes.NewReader(raw))
	if err != nil {
		return fmt.Errorf("create Algorithm request: %w", err)
	}
	request.Header.Set("Content-Type", "application/json")
	response, err := a.Client.Do(request)
	if err != nil {
		return fmt.Errorf("call Algorithm: %w", err)
	}
	defer response.Body.Close()
	responseBody, err := io.ReadAll(io.LimitReader(response.Body, 4<<20))
	if err != nil {
		return fmt.Errorf("read Algorithm response: %w", err)
	}
	if response.StatusCode < 200 || response.StatusCode >= 300 {
		return fmt.Errorf("Algorithm HTTP %d: %s", response.StatusCode, string(responseBody))
	}
	if err := json.Unmarshal(responseBody, result); err != nil {
		return fmt.Errorf("decode Algorithm response: %w", err)
	}
	return nil
}
